### PR TITLE
0.23 JVMism cleanups

### PR DESCRIPTION
**Pull request by armanbilge**
Setting us up for a JVMism free 1.0, either directly or by deprecation.

### CODE REVIEW

File: core/shared/src/main/scala/org/http4s/Message.scala
Comment (armanbilge) [inline core/shared/src/main/scala/org/http4s/Message.scala:412]: 0.23 JS won't be getting this one since it needs an `Async`.

Actually, I will add this upstream to ip4s so in 1.0 we can just ask for an `Dns` capability here instead of `Sync`/`Async`.

https://github.com/Comcast/ip4s/blob/main/jvm/src/main/scala/com/comcast/ip4s/Dns.scala

File: tests/shared/src/test/scala/org/http4s/Ipv4AddressSuite.scala
Comment (armanbilge) [inline tests/shared/src/test/scala/org/http4s/Ipv4AddressSuite.scala:41]: Or else, we should just remove these tests. If we address the deprecation then we are just testing if ip4s works :)

Comment (armanbilge) [reply tests/shared/src/test/scala/org/http4s/Ipv4AddressSuite.scala]: Forgive me, I just deleted them after getting into a twist of `nowarn` and `nowarn` doesn't suppress warnings between platforms.

File: core/shared/src/main/scala/org/http4s/Message.scala
Comment (armanbilge) [reply core/shared/src/main/scala/org/http4s/Message.scala:412]: https://github.com/Comcast/ip4s/pull/325

Comment (rossabaker) [reply core/shared/src/main/scala/org/http4s/Message.scala:412]: I'm missing something.  How is JS not getting this?  This is in shared?

Comment (armanbilge) [reply core/shared/src/main/scala/org/http4s/Message.scala:412]: Scala.js is super weird :(

With Scala.js, you can pretty much _compile_ any code you want, even it's loaded with tons of JVM-only stuff. It's only when you actually try and generate a `.js` file (i.e., the _linking_ stage), that if you used one of these JVMisms it can't find a JavaScript implementation to link to and stops you in your tracks.

_However_, the first part of the linking stage is dead-code-elimination. So, if nobody uses this `remoteHost` method in their Scala.js app, then it will be dead-code-eliminated and the reference to the JVMism will disappear along with all the associated problems. In fact, the reason I didn't catch this earlier must be that our test suite never calls this method, and therefore it is always dead-code-eliminated.

This weirdness is exactly what allowed us to pull off the trick with the syntax shims: although they compile under Scala.js, as long as nobody uses them (or anything that references them) they get dead-code-eliminated before they cause problems.

So what I meant by JS "not getting this" is that because it is currently implemented using a JVMism, although this code will compile in our JS build, nobody will be able to link to it, and thus they won't "get it".

The polite thing to do here, would be to "platform"/split the file, and move this method into JVM-only sources. Because otherwise, some pour soul can accidentally call this method, compile their code (even ship a library!) but its only when somebody tries to _link_ it (presumably because they want to run it) will they find out they've been fooled all along.

IDK if that helps 😅

Comment (armanbilge) [reply core/shared/src/main/scala/org/http4s/Message.scala:412]: Oh also, I think I have an idea for a binary- and source-compatible way to adjust this method so that it _can_ be supported on JVM and JS, after my ip4s PR is released.

Comment (rossabaker) [reply core/shared/src/main/scala/org/http4s/Message.scala:412]: I guess I'm still having a hard time understanding why the Scala.js compiler sees all these signatures it won't be able to link.  If we can so easily add these dead code bombs by having unreferenced-on-Scala.js functionality, it sounds like code coverage becomes more important as a safety device.

Comment (armanbilge) [reply core/shared/src/main/scala/org/http4s/Message.scala:412]: > I guess I'm still having a hard time understanding why the Scala.js compiler sees all these signatures it won't be able to link.

I'll try and find a reference for this, but my understanding is that Scala.js compiles against `.class` files (such as the ones provided in the JDK) but links against `.sjsir` files. And these aren't necessarily consistent with each other (in the case of the standard Java lib: definitely not). I'm sure there are Reasons why it works this way but I don't know.

> it sounds like code coverage becomes more important as a safety device.

Yes. This. And not just in our test suite maybe ... I was only able to catch https://github.com/http4s/http4s/pull/5307 because I published locally and tried building http4s-dom against 0.23sjs. Somehow that issue was masked in our own test suite.

Tons of weird stuff can happen because of screw ups that fly under the radar. Like consider this PR which fixed Scala.js support in cats-effect-testing: https://github.com/typelevel/cats-effect-testing/pull/180. It behaved like it was working and published Scala.js artifacts and everything, although they were completely useless. I only uncovered the problem when I tried to use those artifacts downstream and even then it took me a while to realize what the underlying problem was: https://github.com/typelevel/cats-effect-testing/pull/176#issuecomment-887583137

**Comment by rossabaker** [comment]
Do you want to try to squeeze the Dns in before merging this?  Do we need to wait on that to release 0.23?

**Comment by armanbilge** [comment]
No, the `Dns` can come later. No need to block 0.23 on that.

**Comment by armanbilge** [comment]
But I would still prefer to get https://github.com/typelevel/fs2/pull/2655 out first. What's the timeline for the release train?

File: core/shared/src/main/scala/org/http4s/Message.scala
Comment (armanbilge) [reply core/shared/src/main/scala/org/http4s/Message.scala:412]: Ok, I found the reference I was looking for:

> This will generate .sjsir and .class files for each class in your project (just like for Scala JVM). The .class files are used by the Scala.js compiler for symbol lookup during separate compilation.

https://www.scala-js.org/doc/project/building.html

**Comment by rossabaker** [comment]
0.21 is ready now.  I'm gathering notes for 0.22.  Nothing says they need to go together, but it's frustrating when old versions get shiny things first.

**Comment by armanbilge** [comment]
https://github.com/typelevel/fs2/pull/2655 is on the way.

File: async-http-client/src/main/scala/org/http4s/asynchttpclient/AsyncHttpClient.scala
Comment (armanbilge) [inline async-http-client/src/main/scala/org/http4s/asynchttpclient/AsyncHttpClient.scala:119]: This is a ride-along deprecation that appeared when I bumped fs2.

**Comment by armanbilge** [comment]
I bumped fs2 to a hash so I could make the final changes.

**Comment by armanbilge** [comment]
Seems there's something changed in fs2 since 3.1.3 that causes ember-server to fail tests. Ignoring this PR here, if on `series/0.23` I bump fs2 to `3.1-69-b2a90a1` basically every test fails with:
```
==> X org.http4s.ember.server.EmberServerSuite.server responds to requests  0.015s java.net.BindException: Address already in use
```

**Comment by armanbilge** [comment]
Ok, I tracked it down to https://github.com/typelevel/fs2/pull/2644.

**Comment by rossabaker** [comment]
If a release is imminent, we can probably start rolling 0.21 and 0.22 and pick up 0.23 no later than this week.

**Comment by armanbilge** [comment]
Let me ping @mpilquist, when I asked yesterday it seemed possible but then I found that new regression (which btw I've fixed in https://github.com/typelevel/fs2/pull/2665).

**Comment by armanbilge** [comment]
fs2 is on the way.

**Review by rossabaker — APPROVED** [review]
:+1: on green

**Pull request merged**
